backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
     timeout: i32,
     mut child: Child,
 ) -> crate::error::Result<ExitStatus> {
-    let job_id = job_id.clone();
-    let stderr = child
-        .stderr
-        .take()
-        .expect("child did not have a handle to stdout");
+    let timeout = Duration::from_secs(u64::try_from(timeout).expect("invalid timeout"));
+    let ping_interval = Duration::from_secs(5);
+    let cancel_check_interval = Duration::from_millis(500);
+    let write_logs_delay = Duration::from_millis(500);
 
-    let stdout = child
-        .stdout
-        .take()
-        .expect("child did not have a handle to stdout");
+    let (set_too_many_logs, mut too_many_logs) = watch::channel::<bool>(false);
 
-    let mut reader = BufReader::new(stdout).lines();
-    let mut stderr_reader = BufReader::new(stderr).lines();
-
-    let done = Arc::new(AtomicBool::new(false));
-
-    let done2 = done.clone();
-    let done3 = done.clone();
-    let done4 = done.clone();
-    // Ensure the child process is spawned in the runtime so it can
-    // make progress on its own while we await for any output.
-    let handle = tokio::spawn(async move {
-        let inner_done = done2.clone();
-        let r: Result<ExitStatus, anyhow::Error> = tokio::select! {
-            r = child.wait() => {
-                inner_done.store(true, Ordering::Relaxed);
-                Ok(r?)
+    let output = child_joined_output_stream(&mut child);
+    let job_id = job_id.clone();
+
+    /* the cancellation future is polled on by `wait` while
+     * waiting for the child to exit normally */
+    let cancel_db = db.clone();
+    let cancel_check = async move {
+        let db = cancel_db;
+        let mut interval = repeat(cancel_check_interval).then(sleep).boxed();
+        while let Some(()) = interval.next().await {
+            if sqlx::query_scalar!("SELECT canceled FROM queue WHERE id = $1", job_id)
+                .fetch_optional(&db)
+                .await
+                .map(|v| Some(true) == v)
+                .unwrap_or_else(|err| {
+                    tracing::error!(%job_id, %err, "error checking cancelation");
+                    false
+                })
+            {
+                break;
             }
-            _ = async move {
-                while !done2.load(Ordering::Relaxed) {
-                    tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    };
+
+    /* a future that completes when the child process exits */
+    let wait_db = db.clone();
+    let wait = async move {
```

Comment:
I am not too sure about the name of this: wait_result_or_kill ?

Reply:
`wait_on_child`?

Reply:
:+1: